src/core/trulens/core/utils/python.py

```diff
@@ -1223,85 +1144,55 @@ def delete_singleton_by_name(
                     continue
 
-                del SingletonPerName._instances[k]
-                del SingletonPerName._id_to_name_map[id(v.val)]
+                del SingletonPerNameMeta._singleton_instances[k]
 
-    def delete_singleton(self):
+    @staticmethod
+    def delete_singleton(
+        obj: Type[SingletonPerNameMeta], name: Optional[str] = None
+    ):
         """
         Delete the singleton instance. Can be used for testing to create another
         singleton.
         """
-        id_ = id(self)
-
-        if id_ in SingletonPerName._id_to_name_map:
-            name = SingletonPerName._id_to_name_map[id_]
-            del SingletonPerName._id_to_name_map[id_]
-            del SingletonPerName._instances[(self.__class__.__name__, name)]
+        cls_name = (
+            getattr(obj.__class__, "__name__")
+            if hasattr(obj.__class__, "__name__")
+            else None
+        )
+        k = cls_name, name
+        if k in SingletonPerNameMeta._singleton_instances:
+            del SingletonPerNameMeta._singleton_instances[k]
         else:
-            logger.warning(
-                "Instance %s not found among existing singletons.", self
-            )
-
-
-class Singleton:
-    """Class for creating singleton instances."""
+            logger.warning("Instance %s not found:", obj)
 
-    # Hold singleton instances here.
-    _instances: Dict[Hashable, SingletonInfo[Singleton]] = {}
 
-    def warning(self):
-        """Issue warning that this singleton already exists."""
-
-        k = self.__class__.__name__
-        if k in Singleton._instances:
-            Singleton._instances[k].warning()
-        else:
-            raise RuntimeError(
-                f"Instance of singleton type {k} does not exist."
-            )
+class PydanticSingletonMeta(type(pydantic.BaseModel), SingletonPerNameMeta):
+    """This is the metaclass for creating Pydantic models that are also required to be singletons"""
 
-    def __new__(
-        cls: Type[Singleton],
-        *args,
-        **kwargs,
-    ) -> Singleton:
-        """Create the singleton instance if it doesn't already exist and return it."""
+    pass
 
-        k = cls.__name__
 
-        if k not in cls._instances:
-            logger.debug(
-                "Creating new %s singleton instance.",
-                cls.__name__,
-            )
-            # If exception happens here, the instance should not be added to
-            # _instances.
-            instance = super().__new__(cls)
+class InstanceRefMixin:
+    """Mixin for classes that need to keep track of their instances."""
 
-            info: SingletonInfo = SingletonInfo(val=instance)
-            Singleton._instances[k] = info
-        else:
-            info = Singleton._instances[k]
-
-        obj = info.val
-        assert isinstance(obj, cls)
-        return obj
+    _instance_refs: Dict[
+        Type, weakref.WeakSet[weakref.ReferenceType[InstanceRefMixin]]
```

Comment:
Don't need to stack the refs. Store just the `InstanceRefMixin` inside the WeakSet.